### ISSUE TITLE
Hosting from subpath via reverse proxy

### ISSUE THREAD

**Issue by schmurian** [sbondCo/Watcharr#312]
Hi,

I have another question:
Would it be possible to access Watcharr from a subpath via nginx or similar reverse proxies?

An example configuration could look something like this in nginx:
```
...
location /watcharr/ {
                proxy_pass http://localhost:3080/;
                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Proto $scheme;
        }
...
```

**Comment by IRHM** [comment]
Hi @schmurian, the beta site is hosted behind a reverse proxy using Caddy. I have never tried doing it behind a subpath (nor have I used nginx before), but I will try giving it a go to see if it works.

---

I think I see the problem, the links/redirects won't include the new base path which causes it to break. I'll look into how we can fix that, probably a new `base url` setting.

---

Yep, this is harder than I thought. The only thing I can think of is to manually add the base url (if set) to the start of each url.

I can live with that solution, but it seems very prone to being broken when I inevitably forget to do it.

Going to think some more and probably `just do it` anyways because I'm not that smart.

**Comment by schmurian** [comment]
Please don't stress about this request.
Yes, you're right, I've seen the `base path` being exposed in other docker configuration files, but I remember it can be complex to implement it in a way that it doesn't break things.
I can use a subdomain instead in nginx.

Thank you for looking into it though!

**Comment by christaikobo** [comment]
A potentially easier way to implement this is to write all paths as relative. 

Almost all reverse proxy can strip URL prefix when proxying to upstream. Like in caddy you can simply use "handle_path". If every path is relative then it just works, without the need to specify base_url. 

I have seen other project use this approach.

**Comment by IRHM** [comment]
@terrytw thank you for the suggestion, i hadn't thought of something like that before.

I'd have to have a closer look at some places where we navigate. Off the top of my head, it seems like it might not be possible for links to pages (eg the discover button on the nav links to `/discover`, which works from anywhere in the app, but I don't think theres a way to make that work with a relative link from everywhere. If I am correct in that, would you know a way around it?).

If the above doesn't work, It could be that I revisit the base url solution and just write a helper function that we use in place of direct navigation everywhere.

Been a while since I have looked into this, apologies if I have missed anything.

**Comment by christaikobo** [comment]
Your concern is valid, I thought over it and I think you are right, with a navigation panel on all pages it does make it difficult to use relative path. The project I mentioned which used relative path usually doesn't have it. 

I went over `layout.svelte` and `routes.go` a bit, but since I haven't got the first clue about go, I am totally lost...

If base_url is really needed, then it seems to be quite a bit of work.

**Comment by IRHM** [comment]
Thanks for looking over and confirming @terrytw.

I suppose we will have to go that route then.

**Comment by christaikobo** [comment]
A random thought:

It seems that the frontend is built using vite. Before you got the time to write the helper, is it possible to provide a base docker image for users to build the front end themselves by manually adding `base: '/desired_subpath'` to `vite.config.js`? 

I did it for another project called traccar, which also uses vite to build its frontend. It worked. It seems that most frontend related projects already have subpath in mind.

**Comment by IRHM** [comment]
> A random thought:
> 
> It seems that the frontend is built using vite. Before you got the time to write the helper, is it possible to provide a base docker image for users to build the front end themselves by manually adding `base: '/desired_subpath'` to `vite.config.js`?
> 
> I did it for another project called traccar, which also uses vite to build its frontend. It worked. It seems that most frontend related projects already have subpath in mind.

Unfortunately I don't have much time to test that out at the minute, if you can find the spare time, you could add that property to our vite config and attempt a build with the dockerfile.

It's likely that doing that might not allow any api requests to go through (baseURL var for api is absoluted to "/api", don't think it'll take into account any subpath): https://github.com/sbondCo/Watcharr/blob/565d90dbf042915730badfdc07287171f6e33f25/src/lib/util/api.ts#L26

Out of curiosity is there any particular reason you'd choose to host under a subpath rather than subdomain?

**Comment by christaikobo** [comment]
> > A random thought:
> > It seems that the frontend is built using vite. Before you got the time to write the helper, is it possible to provide a base docker image for users to build the front end themselves by manually adding `base: '/desired_subpath'` to `vite.config.js`?
> > I did it for another project called traccar, which also uses vite to build its frontend. It worked. It seems that most frontend related projects already have subpath in mind.
> 
> Unfortunately I don't have much time to test that out at the minute, if you can find the spare time, you could add that property to our vite config and attempt a build with the dockerfile.
> 
> It's likely that doing that might not allow any api requests to go through (baseURL var for api is absoluted to "/api", don't think it'll take into account any subpath):
> 
> [Watcharr/src/lib/util/api.ts](https://github.com/sbondCo/Watcharr/blob/565d90dbf042915730badfdc07287171f6e33f25/src/lib/util/api.ts#L26)
> 
> Line 26 in [565d90d](/sbondCo/Watcharr/commit/565d90dbf042915730badfdc07287171f6e33f25)
> 
>  : "/api"; 
> Out of curiosity is there any particular reason you'd choose to host under a subpath rather than subdomain?

Thanks for the reply.

I have actually tried just as I commented, and it didn't work....

I noticed the change to API was needed and I made change to `api.ts`, and the api calls were successfully redirected to subpath. 

But the config in vite.config.ts and the svelte config file didn't work somehow, and all the pages were still at root instead of subpath. I think it might be hardcoded somewhere but I am unable to find it. 

As for the reason why I want to host it under subpath, it is mostly due to convenience and privacy. I host everything under 1 subdomain, it is less DNS and certificates to worry about, and by using subpath it is difficult for ISPs (or anyone really) to know how many services I have running on my server.

**Comment by IRHM** [comment]
That is unfortunate @christaikobo, I have no idea where something might be hardcoded. It sounds like it is still something in the build steps that isn't configured right judging by output?

I had never thought of the privacy concern as a reason to use subpaths, thank you for educating me.

**Comment by christaikobo** [comment]
> That is unfortunate [@christaikobo](https://github.com/christaikobo), I have no idea where something might be hardcoded. It sounds like it is still something in the build steps that isn't configured right judging by output?

Yes I think so, somewhere in the code. But it is beyond my capabilities, I tried semi hard spending a couple of hours but unable to figure it out. 

In the long run it is obviously better to have the helper as you mentioned. Even if I made this work, which I failed to do so, it is still a workaround after all. But there must other priorities on the to do list and I understand that, just going to wait patiently.

**Comment by IRHM** [comment]
I did some testing last night, it looks like we set a base path in the [svelte config file](https://github.com/sbondCo/Watcharr/blob/dev/svelte.config.js) (under the `kit` property): https://svelte.dev/docs/kit/configuration#paths (then after a build, it is correctly accessible under the chosen subpath).

I was able to get a small test working under a subpath (with caddy) and got it loading the login screen correctly (we still have to go through each link and add the base).

I was hoping to try getting support for this in our existing docker image, but judging from my testing it doesn't look like this base path can be configured after a build. There doesn't seem to be a lot of resouces online for what I want to do.

In an ideal world, after you update the base path in your server dashboard, the ui server is restarted to use it (without requiring a rebuild).

**Comment by christaikobo** [comment]
> I did some testing last night, it looks like we set a base path in the [svelte config file](https://github.com/sbondCo/Watcharr/blob/dev/svelte.config.js) (under the `kit` property): https://svelte.dev/docs/kit/configuration#paths (then after a build, it is correctly accessible under the chosen subpath).
> 
> I was able to get a small test working under a subpath (with caddy) and got it loading the login screen correctly (we still have to go through each link and add the base).

Weird, I did the same thing and it just didn't work.

```
  kit: {
    paths: {
      base: '/watcharr/',
    },
  },
```

This is what I did. I never correctly landed on login or any other page...I'll try it one more time when I got home. 

> I was hoping to try getting support for this in our existing docker image, but judging from my testing it doesn't look like this base path can be configured after a build. There doesn't seem to be a lot of resouces online for what I want to do.
> 
> In an ideal world, after you update the base path in your server dashboard, the ui server is restarted to use it (without requiring a rebuild).

I think a rebuild is inevitable if we are going to do it this way. 

Related discussion: https://github.com/sveltejs/kit/issues/595

**Comment by IRHM** [comment]
@christaikobo If you remove the trailing slash from `base` it should work (eg just "/watcharr").

I had to edit code for the redirect to login page to work. Going to `/watcharr/login` directly may work as a test.

> I think a rebuild is inevitable if we are going to do it this way.

Very sad, but I think you're right. I believe we are forced to use this `base` parameter, otherwise the sveltekit bootstrap logic itself will attempt to import the app from the wrong path. I don't think theres a way around that.

**Comment by christaikobo** [comment]
> Very sad, but I think you're right. I believe we are forced to use this `base` parameter, otherwise the sveltekit bootstrap logic itself will attempt to import the app from the wrong path. I don't think theres a way around that.

Bright side is that the rebuild is only like 10 seconds. And it most likely will only be run once for the majority of people.

**Comment by IRHM** [comment]
Just wanted to give an update on this, the last I was testing this, it was working on the [subpath branch](https://github.com/sbondCo/Watcharr/tree/subpath).

I realised it comes with security issues because of how we store user tokens in localStorage, being that (I believe) any other app hosted on the same domain would have access to it. I'm not sure if it's a good idea releasing it in that state where tokens could be compromised so easily.

I'm realising localStorage probably isn't ideal for token storage so I may look into changing how that works in the future before continuing with this.

Happy to hear if anyone has any thoughts on this in the meanwhile.
